compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeSceneLayerView.uikit.kt

```diff
@@ -0,0 +1,117 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.scene
+
+import androidx.compose.ui.input.pointer.PointerEventType
+import androidx.compose.ui.window.centroidLocationInView
+import kotlinx.cinterop.CValue
+import kotlinx.cinterop.readValue
+import platform.CoreGraphics.CGPoint
+import platform.CoreGraphics.CGRectZero
+import platform.UIKit.UIEvent
+import platform.UIKit.UITouch
+import platform.UIKit.UIView
+
+/**
+ * A backing ComposeSceneLayer view for each Compose scene layer. Its task is to
+ * handle events that start outside the bounds of the layer content
+ *
+ * @param isInsideInteractionBounds a function that returns true if the given point is inside the
+ * interactable content part of the layer.
+ * @param isInterceptingOutsideEvents a function that returns true if the layer should intercept
+ * events that start outside the bounds of the layer content or should let them pass through.
+ */
+internal class ComposeSceneLayerView(
+    val isInsideInteractionBounds: (point: CValue<CGPoint>) -> Boolean,
+    val isInterceptingOutsideEvents: () -> Boolean
+): UIView(frame = CGRectZero.readValue()) {
+    private var touchesCount: Int = 0
+    private var previousSuccessHitTestTimestamp: Double? = null
+
+    internal var onOutsidePointerEvent: ((
+        eventType: PointerEventType
+    ) -> Unit)? = null
+
+    init {
+        translatesAutoresizingMaskIntoConstraints = false
+        setMultipleTouchEnabled(true)
+    }
+
+    private fun touchStartedOutside(withEvent: UIEvent?) {
+        // hitTest call can happen multiple times for the same touch event, ensure we only send
+        // PointerEventType.Press once using the timestamp.
+        if (previousSuccessHitTestTimestamp != withEvent?.timestamp) {
+            // This workaround needs to send PointerEventType.Press just once
+            previousSuccessHitTestTimestamp = withEvent?.timestamp
+            onOutsidePointerEvent?.invoke(PointerEventType.Press)
+        }
+    }
+
+    override fun touchesBegan(touches: Set<*>, withEvent: UIEvent?) {
+        super.touchesBegan(touches, withEvent)
+
+        touchesCount += touches.size
+    }
+
+    override fun touchesCancelled(touches: Set<*>, withEvent: UIEvent?) {
+        super.touchesCancelled(touches, withEvent)
+
+        touchesCount -= touches.size
+    }
+
+    override fun touchesEnded(touches: Set<*>, withEvent: UIEvent?) {
+        touchesCount -= touches.size
+
+        // It was the last touch in the sequence, calculate the centroid and if it's outside
+        // the bounds, send `onOutsidePointerEvent`. Otherwise, just return.
+        if (touchesCount > 0) {
+            return
+        }
+
+        val location = requireNotNull(
+            touches
+                .map { it as UITouch }
+                .centroidLocationInView(this)
```

Comment:
Adding centroid logic to trigger `onOutsidePointerEvent` seems not valid - it should be triggered only when started outside. Such things shouldn't be changed during refactor, it's out of scope here

Reply:
If two touches were raised at the same frame, what do you want to do? Send an event if at least one touch is outside?

Reply:
It doesn't matter. It should be sent ONLY if the first pointer started outside

Reply:
I think this implementation is out of current task scope. Will be fixed in a separate MR if it caused problesms.